### PR TITLE
Initialize a ByteBuffer from a ByteBufferView

### LINKED ISSUE #946: get a ByteBuffer (slice) from a ByteBufferView
There’s no reason we don’t provide a way to  get a ByteBuffer (slice) from a ByteBufferView.

### PR DESCRIPTION
**Pull request by slashmo**
### Motivation:

As @weissi previously said in (#946), there's no reason not be able to get a `ByteBuffer` from a `ByteBufferView`.

### Modifications:

I added a public initializer to `ByteBuffer` taking a `ByteBufferView` to create a slice of the view's buffer from the same range.

### Result:

`ByteBuffer` will have a public initializer to create a `ByteBuffer` from a `ByteBufferView`.

Closes #946

### CODE REVIEW

File: Sources/NIO/ByteBuffer-views.swift [hunk 1]
```diff
@@ -66,6 +66,14 @@ public struct ByteBufferView: RandomAccessCollection {
             return try body(bytes.bindMemory(to: UInt8.self))
         }
     }
+
+    /// Returns a `ByteBuffer` slice from the same range as the view.
+    ///
+    /// - returns: A `ByteBuffer` slice from the same range as the view or `nil` if the bytes are not readable in the
+    /// initial `ByteBuffer`.
+    public func getByteBuffer() -> ByteBuffer? {
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-views.swift]: I don't think this needs to return `nil`, as it is not possible to obtain a `ByteBufferView` of non-readable bytes anyway. I think we can safely just force-unwrap the return value of `getSlice`.

Comment (slashmo) [reply Sources/NIO/ByteBuffer-views.swift]: Thanks for the suggestion. I changed the method to return a non-optional `ByteBuffer` instead.

File: Sources/NIO/ByteBuffer-views.swift [hunk 2]
```diff
@@ -66,6 +66,13 @@ public struct ByteBufferView: RandomAccessCollection {
             return try body(bytes.bindMemory(to: UInt8.self))
         }
     }
+
+    /// Returns a `ByteBuffer` slice from the same range as the view.
+    ///
+    /// - returns: A `ByteBuffer` slice from the same range as the view.
+    public func getByteBuffer() -> ByteBuffer {
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-views.swift]: I don't think the naming is consistent with Swift(NIO). I could see a couple of options:

1. `public var byteBuffer: ByteBuffer { ... }`, fine as a property because it's constant time
2. `public var buffer: ByteBuffer { ... }`
3. `public func as(Byte)Buffer() -> ByteBuffer`
4.  as a `ByteBuffer` initialiser

```
extension ByteBuffer {
    public init(_ view: ByteBufferView) {
        self = view.buffer.getSlice(....)
    }
}
```

CC @Lukasa ?

File: Tests/NIOTests/ByteBufferTest.swift [hunk 3]
```diff
@@ -1481,6 +1481,16 @@ class ByteBufferTest: XCTestCase {
...
 
+    func testByteBufferViewsReturnByteBuffersOfEqualLength() throws {
+        self.buf.writeString("hello")
+
+        let readableByteBuffer = self.buf.readableBytesView.getByteBuffer()
+        XCTAssertEqual(self.buf, readableByteBuffer)
+
+        let sliceByteBuffer = self.buf.viewBytes(at: 1, length: 3)?.getByteBuffer()
+        XCTAssertEqual("ell", sliceByteBuffer.flatMap { $0.getString(at: 0, length: $0.readableBytes) })
+    }
```
Comment (weissi) [inline Tests/NIOTests/ByteBufferTest.swift]: can we have two more tests for the edge cases:
- the empty case
- where something matches capacity, something like this should work:

```swift
var buffer = self.allocator.buffer(16)
let capacity = buffer.capacity
buffer.writeString(String(repeating: "x", count: capacity))
XCTAssertEqual(capacity, buffer.capacity)
XCTAssertEqual(buffer.viewReadableBytes().getByteBuffer(), buffer)
```

File: Sources/NIO/ByteBuffer-views.swift (hunk 2 above)
Comment (slashmo) [reply Sources/NIO/ByteBuffer-views.swift]: You're right, I forgot that `readableBytesView` is non-optional. 
In that case, I think the initializer would be the most natural choice. Especially if you compare it with `Array` and `ArraySlice`, where you use the `Array` initializer to convert the slice back into an array.

Comment (weissi) [reply Sources/NIO/ByteBuffer-views.swift]: thanks! sounds good to me. @Lukasa happy with this too?

Comment (Lukasa) [reply Sources/NIO/ByteBuffer-views.swift]: Works for me!

Comment (slashmo) [reply Sources/NIO/ByteBuffer-views.swift]: Perfect 👌 I'll change the naming & implement the unit test suggestions from @weissi.

**Comment by weissi** [comment]
@swift-nio-bot test this please
